server/app/views/components/TextFormatter.java

```diff
@@ -1,190 +1,121 @@
-    contentTags.forEach(accordion::addContent);
-    return accordion.getContainer();
+  private static String addIconToLinks(String markdownText) {
+    String closingATag = "</a>";
+    String svgIconString =
+        Icons.svg(Icons.OPEN_IN_NEW)
+            .withClasses("shrink-0", "h-5", "w-auto", "inline", "ml-1", "align-text-top")
+            .toString();
+    return markdownText.replaceAll(closingATag, svgIconString + closingATag);
   }
 
-  private static UlTag buildList(List<String> items) {
-    UlTag listTag = ul().withClasses("list-disc", "mx-8");
-    items.forEach(item -> listTag.with(li().withText(item)));
-    return listTag;
+  private static String addRequiredIndicatorInsidePTag(String markdownText) {
+    int indexOfPTag = markdownText.lastIndexOf("</p>");
+    String stringWithRequiredIndicator = ViewUtils.requiredQuestionIndicator().toString() + "</p>";
+    return markdownText.substring(0, indexOfPTag) + stringWithRequiredIndicator;
+  }
+
+  private static String sanitizeHtml(String markdownText) {
+    PolicyFactory customPolicy =
+        new HtmlPolicyBuilder()
+            .allowElements(
+                "p", "div", "h2", "h3", "h4", "h5", "h6", "a", "ul", "li", "span", "svg", "br",
+                "em", "strong", "code", "path", "pre")
+            // Per accessibility best practices, we want to disallow adding h1 headers to
+            // ensure the page does not have more than one h1 header
+            // https://www.a11yproject.com/posts/how-to-accessible-heading-structure/
+            // This logic changes h1 headers to h2 headers which are still larger than the default
+            // text
+            .allowElements(
+                (String elementName, List<String> attrs) -> {
+                  return "h2";
+                },
+                "h1")
+            .allowWithoutAttributes()
+            .allowAttributes(
+                "class",
+                "target",
+                "xmlns",
+                "fill",
+                "stroke",
+                "stroke-width",
+                "aria-hidden",
+                "viewbox",
+                "d")
+            .globally()
+            .toFactory();
+
+    PolicyFactory policy = customPolicy.and(Sanitizers.LINKS);
+    return policy.sanitize(markdownText, buildHtmlChangeListener(), /*context=*/ null);
+  }
+
+  private static HtmlChangeListener<Object> buildHtmlChangeListener() {
+    return new HtmlChangeListener<Object>() {
+      @Override
+      public void discardedTag(Object ctx, String elementName) {
+        logger.warn(String.format("HTML element: \"%s\" was caught and discarded.", elementName));
+      }
+
+      @Override
+      public void discardedAttributes(Object ctx, String tagName, String... attributeNames) {
+        for (String attribute : attributeNames) {
+          logger.warn(String.format("HTML attribute: \"%s\" was caught and discarded.", attribute));
```

Comment:
Log if we catch elements or attributes that we don't allow

Reply:
(nonblocking) In what circumstances would this logging trigger? If it's rare cool, but if it will spam the logs with warnings Info might be better. The only time I can think of ever wanting to look would being logged is if maybe it's removing a `script` tag I happen to notice, but unless someone on the team has an issue, I probably wouldn't ever look at these.

Reply:
This would trigger if someone tries to pass in markdown that isn't on our "allow" list. I don't see that happening a whole lot unintentionally, so "warn" level felt right. That said, I'm making an assumption and who knows what will happen when actual humans use it! Would you prefer to default to info level and up it to warn later if that feels right or vice versa?

Reply:
In that case let's keep it at warning for now. If it becomes too spammy we can lower it later. Thanks!